File: 0895-maximum-frequency-stack/0895-maximum-frequency-stack.py
class FreqStack:

    def __init__(self):
      self.freq = Counter()
      self.m = defaultdict(list)
      self.max_freq = 0
        

    def push(self, val: int) -> None:
        freq, m = self.freq, self.m
        freq[val] += 1
        
        self.max_freq = max(self.max_freq, freq[val])
        m[freq[val]].append(val)

    def pop(self) -> int:
        freq, m, maxf = self.freq, self.m, self.max_freq
        x = m[maxf].pop()
        if not m[freq[x]]:
            self.max_freq -= 1
        
        freq[x] -= 1
        
        return x
        


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: 0895-maximum-frequency-stack/0895-maximum-frequency-stack.js

var FreqStack = function() {
  this.stack = [];
  this.freq = {};
};

/** 
 * @param {number} val
 * @return {void}
 */
FreqStack.prototype.push = function(val) {
  this.freq[val] ? this.freq[val]++ : this.freq[val] = 1;
  
  const cnt = this.freq[val];
  if (!this.stack[cnt]) this.stack[cnt] = [];
  
  this.stack[cnt].push(val);
};

/**
 * @return {number}
 */
FreqStack.prototype.pop = function() {
  const top = this.stack[this.stack.length - 1];
  const pop = top.pop();
  if (!top.length) this.stack.pop();
  
  this.freq[pop]--;
  
  return pop;
};

/** 
 * Your FreqStack object will be instantiated and called as such:
 * var obj = new FreqStack()
 * obj.push(val)
 * var param_2 = obj.pop()
 */